library/alloc/src/collections/btree/navigate.rs

```diff
@@ -1,12 +1,20 @@
 use core::ptr;
 
 use super::node::{marker, ForceResult::*, Handle, NodeRef};
-use super::unwrap_unchecked;
+
+/// `Option::unwrap` without the promise to panic if the option contains no value.
+/// SAFETY: the caller must ensure that the option contains a value.
+/// Cannot panic if the caller respects safety preconditions.
+#[inline]
+unsafe fn panicless_unwrap_unchecked<T>(val: Option<T>) -> T {
+    val.unwrap_or_else(|| unsafe { core::hint::unreachable_unchecked() })
+}
 
 impl<BorrowType, K, V> Handle<NodeRef<BorrowType, K, V, marker::Leaf>, marker::Edge> {
     /// Given a leaf edge handle, returns [`Result::Ok`] with a handle to the neighboring KV
     /// on the right side, which is either in the same leaf node or in an ancestor node.
     /// If the leaf edge is the last one in the tree, returns [`Result::Err`] with the root node.
+    /// Cannot panic.
```

Comment:
```suggestion
    ///
    /// # Safety
    ///
    /// Cannot panic.
```
Should it be like this to make it more distinct and easier to check? "Cannot panic" after a paragraph might be easy to miss. Since this is internal I think it does not matter much.

Reply:
I general I sure don't know what to do here but some points:

- It's rather distinct from the safety of calling this function, it's a guarantee that this function offers to callers that do not handle panic well (*), and a reminder for implementers.
- I have seen "SAFETY" and "Safety" but not "# Safety".
- Speaking of #, I did see the `#[no_panic]` attribute (or what's it called) in what seems to be [the only attempt that survived](https://crates.io/crates/no-panic) to deal with this formally while diving into [RFC 1736](https://github.com/rust-lang/rfcs/pull/1736) and related. I'm not suggesting to use that crate of course, but maybe reuse the notation as a comment.

(*) But one could state that all functions in node.rs cannot panic. Whenever I make them, by accident or attempt to debug, I'm almost guaranteed a double panic. For instance, because `VacantEntry::insert` first increases the length, then tries to add an element, and if that fails, the drop handler tries to delete that element.

Reply:
> It's rather distinct from the safety of calling this function, it's a guarantee that this function offers to callers that do not handle panic well (*), and a reminder for implementers.

Yes, this sort of stuff are usually let implementors to see.

> I have seen "SAFETY" and "Safety" but not "# Safety".

I see that "SAFETY" is used mainly inside code comments for each `unsafe` use. But here "# Safety" (it also exists somewhere) is used to discuss the safety required when implementing it as part of the doc comments.

Reply:
This shouldn't be made into a safety section (that only makes sense for unsafe functions). However, adding a bit of text saying something like "panicking here leads to UB (or aborts) elsewhere in this module" would be fine.

Reply:
Hmm, repeat that text in all 28 places? For most functions, it's not "in this module" but in the `navigate` module.

I'm quite surprised there is no equivalent to `noexecept` in C++, apparently not even a convention.

How about a very explicit `panicless_` prefix on each name?

Reply:
I don't think that's necessary. We *do* have `#[unwind(aborts)]` but using it throughout here is probably a bad idea - maybe not, not sure.

Reply:
We could add that into our main module doc but I think it is hard for everyone to read all of that.

Reply:
I don't know what either of you mean with "that" and if I did, I probably still wouldn't know how to proceed.

Reply:
@Mark-Simulacrum Can you please share some thoughts to proceed? I don't know what to say.

Reply:
A reminder how we came here: [Mark introduced the `replace` function](https://github.com/rust-lang/rust/pull/68827#discussion_r385134726) to capture a pattern that had been scattered throughout iterator implementations. I got away then without checking or documenting compliance to the no-panic rule. Then Ralf in #73971 asked to add "a comment saying that, and also make sure that indeed next_kv, unwrap_unchecked and next_leaf_edge never panic?" That trickled down to tens of functions receiving a comment in this PR. I'm not at all sure that is what Ralf meant, it could just mean to have a single comment in the new `take_mut` function, the same as Mark wrote (or pasted) in `replace` (as well as removing the `debug_assert`s eager to panic).

I would put either nothing or a "[no_panic]" comment there - it looks less casual, and it's a term you can look up. I haven't found any explanation of "unwind(aborts)", and that (as an attribute) it's used on functions interfacing with C++.

Reply:
> here "# Safety" (it also exists somewhere) is used to discuss the safety required when implementing it as part of the doc comments.

Oh, I see, it's essential to get proper formatting, similar to the examples, such as [NonNull::new_unchecked])https://doc.rust-lang.org/std/ptr/struct.NonNull.html#new_unchecked.v), with bigger blocks recently added. It's not important for all the private modules, but I better get used to doing it properly.